apps/remote_control/lib/lexical/remote_control/code_mod/ast/aliases.ex

```diff
@@ -0,0 +1,199 @@
+    # def func(...)
+    defp do_update(%__MODULE__{} = state, {definition, metadata, _body})
+         when definition in @definitions do
+      push_scope(state, metadata)
+    end
+
+    # A simple alias: alias Foo.Bar
+    defp do_update(%__MODULE__{} = state, {:alias, _metadata, [{:__aliases__, _, source}]}) do
+      dest = List.last(source)
+      put_alias(state, source, dest)
+    end
+
+    # An alias with a specified name: alias Foo.Bar, as: FooBar
+    defp do_update(
+           %__MODULE__{} = state,
+           {:alias, _metadata, [{:__aliases__, _, source}, [as: {:__aliases__, _, [dest]}]]}
+         ) do
+      put_alias(state, source, dest)
+    end
+
+    # A multiple alias: alias Foo.Bar.{First, Second, Third.Fourth}
+    defp do_update(
+           %__MODULE__{} = state,
+           {:alias, _, [{{:., _, [{:__aliases__, _, source_alias}, :{}]}, _, destinations}]}
+         ) do
+      Enum.reduce(destinations, state, fn {:__aliases__, _, destination_alias}, state ->
+        source = source_alias ++ destination_alias
+        dest = List.last(source)
+        put_alias(state, source, dest)
+      end)
+    end
+
+    defp do_update(%__MODULE__{} = state, _elem) do
+      state
+    end
+
+    defp put_alias(%__MODULE__{} = state, source, dest) do
+      %Scope{} = scope = current_scope(state)
+      [first | rest] = source
+
+      # This allows a pre-existing alias to define another alias like
+      # alias Foo.Bar.Baz
+      # alias Baz.Quux
+      source =
+        case state.aliases do
+          %{^first => dest_alias} -> Module.split(dest_alias.source) ++ rest
+          _ -> source
+        end
+
+      new_alias = Alias.new(ensure_alias(source), ensure_alias(dest), scope.ref)
+      %__MODULE__{state | aliases: Map.put(state.aliases, new_alias.dest, new_alias)}
+    end
+
+    defp current_scope(%__MODULE__{scopes: [current | _]}) do
+      current
+    end
+
+    defp ensure_alias(alias_list) when is_list(alias_list) do
+      Module.concat(alias_list)
+    end
+
+    defp ensure_alias(alias_atom) when is_atom(alias_atom) do
+      alias_atom
+    end
+
+    defp push_scope(%__MODULE__{} = state, metadata, on_exit \\ &Function.identity/1) do
+      end_position = {get_in(metadata, [:end, :line]), get_in(metadata, [:end, :column])}
+      %__MODULE__{state | scopes: [Scope.new(end_position, on_exit) | state.scopes]}
+    end
+
+    defp maybe_pop_scope(%__MODULE__{} = state, {_, metadata, _}) do
+      with {:ok, current_line} <- Keyword.fetch(metadata, :line),
+           {:ok, current_column} <- Keyword.fetch(metadata, :column),
+           [current_scope | scopes] <- state.scopes,
+           true <- Scope.ended?(current_scope, {current_line, current_column}) do
+        aliases =
+          Map.reject(state.aliases, fn {_, alias_struct} ->
+            alias_struct.ref == current_scope.ref
+          end)
+
+        %__MODULE__{state | scopes: scopes, aliases: aliases}
+        |> current_scope.on_exit.()
```

Comment:
don't quite understand why `Function.identity/1` here.

Reply:
it's the default `on_exit` callback. It just returns what's been passed in, i.e. it does nothing.